text/0000-cargo-prepublish.md

```diff
@@ -0,0 +1,395 @@
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
+
+[dependencies]
+xml-rs = "0.9.2"
+```
+
+For `servo`, we also need to record the prepublication, but don't need to modify
+or introduce any `xml-rs` dependencies; it's enough to be using the fork of
+`foo`, which we would be anyway:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.9.2" }
+
+[dependencies]
+foo = { git = "https://github.com/aturon/foo", branch = "fix-xml" }
+```
+
+With this setup:
+
+- When compiling `foo`, Cargo will resolve the `xml-rs` dependency to `0.9.2`, and
+retrieve the source from the specified git branch.
+
+- When compiling `servo`, Cargo will again resolve *two* versions of `xml-rs`,
+this time `0.9.2` and `0.8.0`, and for the former it will use the source from
+the git branch.
+
+The Cargo.toml files that needed to be changed here span from the crate that
+actually cares about the new version (`foo`) upward to the root of the crate we
+want to do integration testing for (`servo`); no sibling crates needed to be
+changed.
+
+Once `xml-rs` version `0.9.2` is actually published, we can remove the
+`[augment]` sections, and Cargo will warn us that this needs to be done.
+
+### Scenario: prepublishing a new major version
+
+What happens if `foo` instead needs to make a breaking change to `xml-rs`? The
+workflow is identical. For `foo`:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.10.0" }
+
+[dependencies]
+xml-rs = "0.10.0"
+```
+
+For `servo`:
+
+```toml
+[augment.crates-io]
+xml-rs = { git = "https://github.com/aturon/xml-rs", branch = "0.10.0" }
+
+[dependencies]
+foo = { git = "https://github.com/aturon/foo", branch = "fix-xml" }
+```
+
+However, when we compile, we'll now get *three* versions of `xml-rs`: `0.8.0`,
+`0.9.1` (retained from the previous lockfile), and `0.10.0`. Assuming that
+`xml-rs` is a public dependency used to communicate between `foo` and `bar` this
+will result in a compilation error, since they are using distinct versions of
+`xml-rs`. To fix that, we'll need to update `bar` to also use the new, `0.10.0`
+prepublication version of `xml-rs`.
+
+(Note that a
+[private dependency](https://github.com/rust-lang/cargo/issues/2064) distinction
+would help catch this issue at the Cargo level and give a maximally informative
+error message).
+
+## Impact on `Cargo.lock`
+
+Usage of `[augment]` will perform backwards-incompatible modifications to
+`Cargo.lock`, meaning that usage of `[augment]` will prevent previous versions
+of Cargo from interpreting the lock file. Cargo will unconditionally resolve all
+entries in the `[augment]` section to precise dependencies, encoding them all in
+the lock file whether they're used or not.
+
+Dependencies formed on crates listed in `[augment]` will then be listed directly
+in Cargo.lock, eschewing the actual dependency entirely. For example let's say
```

Comment:
You are correct yeah. I see "eschewing the actual dependency" in terms of "xml-rs 0.10.0 from crates.io" is not listed in the lock file (because it doesn't exist). In any case I'll see if I can update the wording here to be more clear.